did/ion/request.go

```diff
@@ -0,0 +1,302 @@
+			return fmt.Errorf("public key<%s> is duplicated", publicKey.ID)
+		}
+
+		if len(publicKey.ID) > maxIDLength {
+			return fmt.Errorf("public key<%s> id is too long", publicKey.ID)
+		}
+
+		publicKeys[publicKey.ID] = publicKey
+	}
+
+	// check if services to remove are valid
+	for _, serviceID := range s.ServiceIDsToRemove {
+		if _, ok := services[serviceID]; ok {
+			return fmt.Errorf("service<%s> added and removed in same request", serviceID)
+		}
+
+		if len(serviceID) > maxIDLength {
+			return fmt.Errorf("service<%s> id is too long", serviceID)
+		}
+	}
+
+	// check if public keys to remove are valid
+	for _, publicKeyID := range s.PublicKeyIDsToRemove {
+		if _, ok := publicKeys[publicKeyID]; ok {
+			return fmt.Errorf("public key<%s> added and removed in same request", publicKeyID)
+		}
+
+		if len(publicKeyID) > maxIDLength {
+			return fmt.Errorf("public key<%s> id is too long", publicKeyID)
+		}
+	}
+	return nil
+}
+
+// NewUpdateRequest creates a new update request https://identity.foundation/sidetree/spec/#update
+func NewUpdateRequest(didSuffix string, updateKey, nextUpdateKey crypto.PublicKeyJWK, signer BTCSignerVerifier, stateChange StateChange) (*UpdateRequest, error) {
+	if err := stateChange.IsValid(); err != nil {
+		return nil, err
+	}
+
+	// construct update patches
+	var patches []any
+
+	// services to add
+	if len(stateChange.ServicesToAdd) > 0 {
+		addServicesPatch := AddServicesAction{
+			Action:   AddServices,
+			Services: stateChange.ServicesToAdd,
+		}
+		patches = append(patches, addServicesPatch)
+	}
+
+	// services to remove
+	if len(stateChange.ServiceIDsToRemove) > 0 {
+		removeServicesPatch := RemoveServicesAction{
+			Action: RemoveServices,
+			IDs:    stateChange.ServiceIDsToRemove,
+		}
+		patches = append(patches, removeServicesPatch)
+	}
+
+	// public keys to add
+	if len(stateChange.PublicKeysToAdd) > 0 {
+		addPublicKeysPatch := AddPublicKeysAction{
+			Action:     AddPublicKeys,
+			PublicKeys: stateChange.PublicKeysToAdd,
+		}
+		patches = append(patches, addPublicKeysPatch)
+	}
+
+	// public keys to remove
+	if len(stateChange.PublicKeyIDsToRemove) > 0 {
+		removePublicKeysPatch := RemovePublicKeysAction{
+			Action: RemovePublicKeys,
+			IDs:    stateChange.PublicKeyIDsToRemove,
+		}
+		patches = append(patches, removePublicKeysPatch)
+	}
+
+	// prepare reveal value
+	revealValue, _, err := Commit(updateKey)
+	if err != nil {
+		return nil, err
+	}
+
+	// prepare delta
+	_, nextUpdateCommitment, err := Commit(nextUpdateKey)
+	if err != nil {
+		return nil, err
+	}
+	delta := Delta{
+		UpdateCommitment: nextUpdateCommitment,
+		Patches:          patches,
+	}
+	deltaCanonical, err := CanonicalizeAny(delta)
+	if err != nil {
+		return nil, err
+	}
+	deltaHash, err := HashEncode(deltaCanonical)
+	if err != nil {
+		return nil, err
+	}
+
+	// prepare signed data
+	toBeSigned := struct {
+		UpdateKey crypto.PublicKeyJWK `json:"updateKey"`
+		DeltaHash string              `json:"deltaHash"`
+	}{
```

Comment:
Curious why an anonymous struct here vs defining it in `model.go` ?

Reply:
ah was all anonymous first, forgot to update this